Added Pythia-based Rhadron decays during Geant simulation

ctdax opened the pull request:
#### PR description:

<!-- Please replace this text with a description of the feature proposed or problem addressed, specifying:
  - what changes are expected in the output if any, 
  - what other PRs or externals it depends upon if any,
  - link to any additional material useful to provide a documentation for this PR (slides, JIRA tickets, related pull requestes, hypernews, TWiki or Indico pages)  -->

This pull request introduces the ability to decay R-hadrons during simulation using a side instance of Pythia8. A generator is required to decay the underlying SUSY particle to hadronize the outgoing partons. This will allow users to simulate detector interactions of long-lived R-hadrons with proper decays for the first time.

[See the talk given here.](https://indico.cern.ch/event/1613012/#76-r-hadron-decays)

##### How to use it
- Create a SLHA decay table that includes the decays and widths of the underlying SUSY particle in the R-hadron, e.g. the gluino or the stop.
- Pass the SLHA decay table to `SLHAFileForPythia8` inside of process.generator or process.g4SimHits in the python config file. Optionally, commands can be passed directly to the instance of pythia that handles the R-hadron decay via `RhadronPythiaDecayerCommandFile`, e.g.
 
```python
process.generator = cms.EDFilter("Pythia8ConcurrentGeneratorFilter",
    SLHAFileForPythia8 = cms.string('SimG4Core/CustomPhysics/data/TESTDECAY_GLUINO1800_STOP500.txt'),
    RhadronPythiaDecayerCommandFile = cms.untracked.string('SimG4Core/CustomPhysics/data/RhadronPythiaDecayerCommands.txt'),
...
)
```
- Make sure the event generator doesn't decay the SUSY particle before it is handed to Geant for simulation. If the event generator is pythia, you can force the SUSY particle to not decay via `pythia8CommonSettings` in the python config file:

```python
process.generator = cms.EDFilter("Pythia8ConcurrentGeneratorFilter",
    PythiaParameters = cms.PSet(
        pythia8CommonSettings = cms.vstring(
            '1000021:mayDecay = off'
        )
    )
)
```

##### Explanation of changes
- **RHadronPythiaDecayer.cc/RHadronPythiaDecayer.h:** Utilizes G4Decay::DecayIt() to decay the R-hadron. Because it is defined as an external decayer, when G4Decay::DecayIt() calls ImportDecayProducts() it calls the function inside of RHadronPythiaDecayer. This file strips the incoming R-hadron down to its constituents, decays the underlying SUSY particle, then lets Pythia handle the hadronization. The decay products are returned to G4Decay::DecayIt() for them to be simulated in Geant.
- **SimG4Core/CustomPhysics/python/Exotica_HSCP_SIM_cfi.py:** Handles the inputs `SLHAFileForPythia8` and `RhadronPythiaDecayerCommandFile`. Automatically produces and keeps the GenPlusSimParticle collection.
- **SimG4Core/CustomPhysics/src/CustomParticleFactory.cc:** Forces the lifetime of R-hadrons to match the lifetime of their underlying SUSY particle, which is read in from the SLHA table. The modification to Line298 `if ((ToLower(line).find("block") < line.npos) || (line == "#")) {` forces the skipping of commented lines in the SLHA text file.
- **SimG4Core/CustomPhysics/src/CustomPhysicsList.cc:** Sets the R-hadron decay process to be the external pythia decayer. 
- **SimG4Core/CustomPhysics/data/RhadronPythiaDecayerCommands.txt:** A new .txt file that contains default commands for the Pythia instance that decays the Rhadrons.
- **SimG4Core/CustomPhysics/python/CustomPhysics_cfi.py:** Adds the `RhadronPythiaDecayerCommandFile` parameter to the config file.
- **PhysicsTools/HepMCCandAlgos/plugins/GenPlusSimParticleProducer.cc‎:** A special case if statement needed to be added to the producer to add the R-hadron decay products to the collection in the scenario where the R-hadron has a small ctau. Because of how the producer works, the decay products of the R-hadron would only be added to the collection if the R-hadron first flipped types before decaying. This does not often happen when the ctau of the R-hadron is small.
- **SimG4Core/CustomPhysics/BuildFile.xml‎:** Add pythia interface for the RHadronPythiaDecayer.

##### Confirmation of 4-momentum conservation at the decay
A histogram plotting the difference in 4-momentum between the R-hadron and its decay products in 50 events is shown below. These are the 4-momenta in the $p_T$, $\eta$, $\phi$ basis after the decay products are handed back to Geant. All entries are consistent with 0 difference except for some floating-point error.

<img width="2176" height="1166" alt="4pDifferenceAtDecay" src="https://github.com/user-attachments/assets/85d4c379-9d14-4386-9218-85dc2a8cc7c6" />


#### PR validation:

<!-- Please replace this text with a description of which tests have been performed to verify the correctness of the PR, including the eventual addition of new code for testing like unit tests, test configurations, additions or updates to the runTheMatrix test workflows -->

All tests below ran successfully.
```bash
cd src
scram b distclean 
git cms-checkdeps -a -A
scram b -j 8
scram b runtests use-ibeos
runTheMatrix.py -l limited -i all --ibeos
```

Before submitting your pull requests, make sure you followed this checklist:
- verify that the PR is really intended for the chosen branch
- verify that changes follow [CMS Naming, Coding, And Style Rules](http://cms-sw.github.io/cms_coding_rules.html)
- verify that the PR passes the basic test procedure suggested in the [CMSSW PR instructions](https://cms-sw.github.io/PRWorkflow.html)

<!-- Please delete the text above after you verified all points of the checklist  -->

ctdax committed:
Added Pythia-based Rhadron decays during Geant simulation

ctdax committed:
Code format and checks

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-49414/46869

cmsbuild commented:
A new Pull Request was created by  @ctdax for master.

It involves the following packages:

- SimG4Core/CustomPhysics (**simulation**)


@civanch, @cmsbuild, @kpedro88, @mdhildreth can you please review it and eventually sign? Thanks.
@bsunanda, @fabiocos, @makortel, @martinamalberti, @rovere, @slomeo this is something you requested to watch as well.
@ftenchini, @mandrenguyen, @sextonkennedy you are the release manager for this.

cms-bot commands are listed <a href="http://cms-sw.github.io/cms-bot-cmssw-cmds.html">here</a>

ctdax commented:
> @ctdax thanks for this contribution! It will require some review. Are you available to present this work at Friday's SIM meeting?

Of course! Where can I find the indico for the meeting?

kpedro88 commented on SimG4Core/CustomPhysics/src/CustomPhysicsList.cc:
```diff
@@ -62,6 +64,13 @@ void CustomPhysicsList::ConstructProcess() {
+  G4Decay* pythiaDecayProcess = new RHadronPythiaDecayer(myConfig);
+  G4VExtDecayer* extDecayer = dynamic_cast<G4VExtDecayer*>(pythiaDecayProcess);
+  pythiaDecayProcess->SetExtDecayer(
+      extDecayer);  // Set the external decayer to itself. Seems redundant but is necessary as far as I can tell. Without doing this, RHadronPythiaDecayer::ImportDecayProducts() will not be called.
```
comment on separate line

kpedro88 commented on SimG4Core/CustomPhysics/src/RHadronPythiaDecayer.cc:
```diff
@@ -0,0 +1,351 @@
+}
+
+void RHadronPythiaDecayer::RHadronToConstituents(Pythia8::Event& event) {
+  // This code is very similar to Pythia8::RHadrons::decay(). Unfortunately, it is not possible in this scenario to use Pythia8::RHadrons::decay().
```
Yes, unfortunately `nRHad` is a private member in `Pythia8::RHadrons`. If it were protected, one could make an inherited class and modify its value manually. We could propose to change this in the future.

ctdax commented:
@kpedro88 we decided that we would like to compute GEN-level objects using the decay info. 

Should the new HepMCProduct be a shallow copy or a deep copy of the original HepMCProduct?

ctdax committed:
Addressed PR comments. RHDecayTracer produces a new GenParticleCollection. Removed change to CustomPhysicsList ordering.

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-49414/47532

Code check has found code style and quality issues which could be resolved by applying following patch(s)

- **code-format**:
https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-49414/47532/code-format.patch
e.g. `curl -k https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-49414/47532/code-format.patch | patch -p1`
You can also run `scram build code-format` to apply code format directly

ctdax committed:
code format

kpedro88 commented on SimG4Core/CustomPhysics/plugins/RHDecayTracer.cc:
```diff
@@ -1,83 +1,212 @@
+  // Add R-hadrons from the SimTrack collection to the new collection
+  for (const auto& simTrack : *simTracks) {
+    int pdgId = std::abs(simTrack.type());
+    if (pdgId >= 1000600 && pdgId <= 1999999) {
```
why are the 1000021 and 1000006 not included here?

kpedro88 commented on SimG4Core/CustomPhysics/src/CustomPhysicsList.cc:
```diff
@@ -109,20 +104,26 @@ void CustomPhysicsList::ConstructProcess() {
+            particle->GetParticleType() == "sbaryon") && particle->GetPDGStable() == false) {
+          if (!extRHadronDecayerSet) {
+            // Set the pythia decayer for Rhadrons if they are unstable
+            pythiaDecayProcess = new RHadronPythiaDecayer(myConfig);
```
is this memory subsequently managed by Geant? (I can never keep track of what it does or doesn't delete automatically)

ctdax commented:
> @ctdax I think this is an improvement over the initial version. In addition to the technical comments above, one crucial piece is still missing: actually updating the Python sequences to use the new augmented GenParticles collection downstream.

To do this, do I need to update [SimG4Core_EventContent_cff.py](https://github.com/cms-sw/cmssw/blob/a8a0be07cf733c297800282bffcd4749450f88a6/SimG4Core/Configuration/python/SimG4Core_EventContent_cff.py) to keep the new GenParticle collection in the different steps? Or do I need to do something else/something more?

kpedro88 commented:
I had missed this detail. The same principles apply, except for `genParticles` instead of `generatorSmeared`. You will have to write an equivalent of `GeneratorSmearedProducer` e.g. `GenParticleExtraProducer`.

Is it important to remake the derived generator products (GenJets, GenMET, etc.) to be consistent with the new collection? Then we would need to rerun the relevant modules (using [fixGenInfoTask](https://github.com/cms-sw/cmssw/blob/20a7dd3a9341984900e281daf54f3343c9596af2/Configuration/StandardSequences/python/Generator_cff.py#L67), as is currently done in the DIGI step as noted above).

ctdax committed:
Made RHDecayTracer PDG coverage customizable. RHDecayTracer collection now propagates to miniAOD. Addressed other PR comments

ctdax committed:
Code format and checks

ctdax commented:
In the latest commit I addressed all of the comments above and also updated [‎PhysicsTools/PatAlgos/python/slimming/MicroEventContent_cff.py‎](https://github.com/cms-sw/cmssw/pull/49414/changes#diff-d0039c91d7a9b3547d78b5bf135fa3cf67f6f6e1016f0efff7ee76bbb3241bb0)[‎PhysicsTools/PatAlgos/python/slimming/MicroEventContent_cff.py‎](https://github.com/cms-sw/cmssw/pull/49414/changes#diff-d0039c91d7a9b3547d78b5bf135fa3cf67f6f6e1016f0efff7ee76bbb3241bb0)

 and 

[‎SimG4Core/Configuration/python/SimG4Core_EventContent_cff.py‎](https://github.com/cms-sw/cmssw/pull/49414/changes#diff-7d0cae702582f2c28747eca3d963cf57461b9a47deed8c330a379323a3821b9b)[‎SimG4Core/Configuration/python/SimG4Core_EventContent_cff.py‎](https://github.com/cms-sw/cmssw/pull/49414/changes#diff-7d0cae702582f2c28747eca3d963cf57461b9a47deed8c330a379323a3821b9b) 

to keep the new `recoGenParticles_RHDecayTracer_RHadronDecay_SIM` all the way to miniAOD.

ctdax committed:
Removed RHDecayTracer. Updated Exotica_HSCP_SIM_cfi.py to automatically create GenPlusSim if HSCP is unstable

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-49414/48640

ERROR: Build errors found during clang-tidy run.
```
src/PhysicsTools/HepMCCandAlgos/plugins/GenPlusSimParticleProducer.cc:299:13: error: expected '(' after 'if' [clang-diagnostic-error]
  299 |     else if {
      |             ^
Suppressed 1156 warnings (1152 in non-user code, 4 NOLINT).
--
gmake: *** [config/SCRAM/GMake/Makefile.coderules:129: code-checks] Error 2
gmake: *** [There are compilation/build errors. Please see the detail log above.] Error 2
```

ctdax committed:
Removed changes to MicroEventContent and SimG4Core_EventContent

ctdax committed:
Fixed silly else if issue

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-49414/48642

- Found files with invalid states:
  - SimG4Core/CustomPhysics/plugins/RHDecayTracer.cc:
    - Added: eaa1f415bb78ab522a4cb5fb37eb65b30502fe9d
    - Modified: 52c714b37d06077b46f7a68ef64d3c2a0b7e3c5b, aaf916d5ccc05aa9e9975be6dc095cd1a418cdbb, 6c1010a73df7385cb26c96ec4e437c1e7d90c76c, 39bfe95fab2b4ba813fceb0ddd35623376d20582, cfd5054e4a87a5cb19c3f67bcc91092ba5d9f2d9
    - Deleted: 802eb0a84d0126a9cd3ba8db3bbcc9216f6b0ac3
  - SimG4Core/CustomPhysics/interface/RHDecayTracer.h:
    - Added: eaa1f415bb78ab522a4cb5fb37eb65b30502fe9d
    - Modified: 52c714b37d06077b46f7a68ef64d3c2a0b7e3c5b, aaf916d5ccc05aa9e9975be6dc095cd1a418cdbb, 6c1010a73df7385cb26c96ec4e437c1e7d90c76c, 39bfe95fab2b4ba813fceb0ddd35623376d20582, cfd5054e4a87a5cb19c3f67bcc91092ba5d9f2d9
    - Deleted: 802eb0a84d0126a9cd3ba8db3bbcc9216f6b0ac3
  - SimG4Core/CustomPhysics/interface/RHadronPythiaDecayDataManager.h:
    - Added: eaa1f415bb78ab522a4cb5fb37eb65b30502fe9d
    - Modified: 52c714b37d06077b46f7a68ef64d3c2a0b7e3c5b, aaf916d5ccc05aa9e9975be6dc095cd1a418cdbb, 6c1010a73df7385cb26c96ec4e437c1e7d90c76c
    - Deleted: 802eb0a84d0126a9cd3ba8db3bbcc9216f6b0ac3

Code check has found code style and quality issues which could be resolved by applying following patch(s)

- **code-format**:
https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-49414/48642/code-format.patch
e.g. `curl -k https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-49414/48642/code-format.patch | patch -p1`
You can also run `scram build code-format` to apply code format directly

ctdax committed:
Code format

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-49414/48643

- Found files with invalid states:
  - SimG4Core/CustomPhysics/plugins/RHDecayTracer.cc:
    - Added: eaa1f415bb78ab522a4cb5fb37eb65b30502fe9d
    - Modified: 52c714b37d06077b46f7a68ef64d3c2a0b7e3c5b, aaf916d5ccc05aa9e9975be6dc095cd1a418cdbb, 6c1010a73df7385cb26c96ec4e437c1e7d90c76c, 39bfe95fab2b4ba813fceb0ddd35623376d20582, cfd5054e4a87a5cb19c3f67bcc91092ba5d9f2d9
    - Deleted: 802eb0a84d0126a9cd3ba8db3bbcc9216f6b0ac3
  - SimG4Core/CustomPhysics/interface/RHDecayTracer.h:
    - Added: eaa1f415bb78ab522a4cb5fb37eb65b30502fe9d
    - Modified: 52c714b37d06077b46f7a68ef64d3c2a0b7e3c5b, aaf916d5ccc05aa9e9975be6dc095cd1a418cdbb, 6c1010a73df7385cb26c96ec4e437c1e7d90c76c, 39bfe95fab2b4ba813fceb0ddd35623376d20582, cfd5054e4a87a5cb19c3f67bcc91092ba5d9f2d9
    - Deleted: 802eb0a84d0126a9cd3ba8db3bbcc9216f6b0ac3
  - SimG4Core/CustomPhysics/interface/RHadronPythiaDecayDataManager.h:
    - Added: eaa1f415bb78ab522a4cb5fb37eb65b30502fe9d
    - Modified: 52c714b37d06077b46f7a68ef64d3c2a0b7e3c5b, aaf916d5ccc05aa9e9975be6dc095cd1a418cdbb, 6c1010a73df7385cb26c96ec4e437c1e7d90c76c
    - Deleted: 802eb0a84d0126a9cd3ba8db3bbcc9216f6b0ac3

cmsbuild commented:
Pull request #49414 was updated. @civanch, @cmsbuild, @kpedro88, @mdhildreth, @tvami can you please check and sign again.

ctdax committed:
Removed if unstable HSCP statement. Create GenPlusSimCollection for all HSCPs

tvami commented:
@cmsbuild please test
